File: regression-tests/sparktkregtests/testcases/models/linear_regression_test.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

"""Tests Linear Regression Model against known values"""
import unittest
from sparktkregtests.lib import sparktk_test


class LinearRegression(sparktk_test.SparkTKTestCase):

    def setUp(self):
        """Build test frame"""
        super(LinearRegression, self).setUp()
        dataset = self.get_file("linear_regression_gen.csv")
        schema = [("c1", float),
                  ("c2", float),
                  ("c3", float),
                  ("c4", float),
                  ("label", float)]

        self.frame = self.context.frame.import_csv(
            dataset, schema=schema)

    @unittest.skip("no publish") 
    def test_model_publish(self):
        """Test publishing a linear regression model"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'])
        model_path = model.publish()
        self.assertIn("hdfs", model_path)
        self.assertIn("tar", model_path)

    def test_model_test(self):
        """Test test functionality"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'])
        output = model.test(self.frame, "label")
        self.assertAlmostEqual(
            model.mean_squared_error, output.mean_squared_error)
        self.assertAlmostEqual(
            model.root_mean_squared_error, output.root_mean_squared_error)
        self.assertAlmostEqual(
            model.mean_absolute_error, output.mean_absolute_error)
        self.assertAlmostEqual(
            model.explained_variance, output.explained_variance)

    def test_model_predict_output(self):
        """Test output format of predict"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'])
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_elastic_net(self):
        """Test elastic net argument"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
            elastic_net_parameter=0.3)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_fix_intercept(self):
        """Test fix intercept argument"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
            fit_intercept=False)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_max_iterations(self):
        """Test max iterations argument"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
                          max_iterations=70)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_reg_param(self):
        """Test regularization parameter argument"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
                          reg_param=0.000000002)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_standardization(self):
        """Test test non-standardized data"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
                          standardization=False)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def test_model_tolerance(self):
        """Test test a different model tolerance"""
        model = self.context.models.regression.linear_regression.train(
            self.frame, "label", ['c1', 'c2', 'c3', 'c4'],
                          tolerance=0.0000000000000000001)
        predict = model.predict(self.frame, ['c1', 'c2', 'c3', 'c4'])
        self._validate_results(model, predict)

    def _validate_results(self, res, predict):
        # validate dictionary entries, weights, and predict results
        self.assertAlmostEqual(res.mean_absolute_error, 0.0)
        self.assertAlmostEqual(res.root_mean_squared_error, 0.0)
        self.assertAlmostEqual(res.mean_squared_error, 0.0)
        self.assertAlmostEqual(res.intercept, 0.0)
        self.assertEqual(res.value_column, "label")
        self.assertItemsEqual(
            res.observation_columns, ['c1', 'c2', 'c3', 'c4'])
        self.assertLess(res.iterations, 150)

        for (i, j) in zip([0.5, -0.7, -0.24, 0.4], res.weights):
            self.assertAlmostEqual(i, j, places=4)

        pd_res = predict.to_pandas(predict.count())
        for index, row in pd_res.iterrows():
            self.assertAlmostEqual(row["label"], row["predicted_value"], places=4)


if __name__ == '__main__':
    unittest.main()

File: sparktk-core/src/test/scala/org/trustedanalytics/sparktk/models/regression/linear_regression/LinearRegressionTest.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.models.regression.linear_regression

import org.apache.spark.sql.Row
import org.apache.spark.sql.catalyst.expressions.GenericRow
import org.scalatest.Matchers
import org.trustedanalytics.sparktk.frame.{ Frame, DataTypes, Column, FrameSchema }
import org.trustedanalytics.sparktk.models.regression.RegressionTestMetrics
import org.trustedanalytics.sparktk.testutils.TestingSparkContextWordSpec

class LinearRegressionTest extends TestingSparkContextWordSpec with Matchers {

  val labeledPoint: Array[Row] = Array(
    new GenericRow(Array[Any](1.0, 1.0)),
    new GenericRow(Array[Any](2.0, 2.0)),
    new GenericRow(Array[Any](3.0, 3.0)),
    new GenericRow(Array[Any](4.0, 4.0)))
  val schema = new FrameSchema(List(Column("label", DataTypes.float64), Column("obs1", DataTypes.float64)))

  "Linear Regression" should {
    "create a linear regression model" in {

      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val model = LinearRegressionModel.train(frame, "label", List("obs1"))

      model shouldBe a[LinearRegressionModel]
      model.valueColumn should equal("label")
      model.observationColumnsTrain should equal(List("obs1"))
      model.intercept should equal(0.0)
      model.weights should equal(List(1.0))
      model.r2 should equal(1.0)
      model.iterations should equal(1)
    }

    "test should return a regression metric" in {

      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val model = LinearRegressionModel.train(frame, "label", List("obs1"))
      val metrics = model.test(frame, "label", None)

      metrics shouldBe a[RegressionTestMetrics]
      metrics.r2 should equal(1.0)
      metrics.explainedVarianceScore should equal(1.0)
      metrics.meanSquaredError should equal(0)
    }

    "predict the results of linear regression correctly" in {

      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val model = LinearRegressionModel.train(frame, "label", List("obs1"))

      val predictFrame = model.predict(frame, None)
      predictFrame.schema.columns should equal(List(Column("label", DataTypes.float64), Column("obs1", DataTypes.float64), Column("predicted_value", DataTypes.float64)))
      predictFrame.rdd.toArray.toList should equal(List(
        new GenericRow(Array[Any](1.0, 1.0, 1.0)),
        new GenericRow(Array[Any](2.0, 2.0, 2.0)),
        new GenericRow(Array[Any](3.0, 3.0, 3.0)),
        new GenericRow(Array[Any](4.0, 4.0, 4.0))))
    }
    "throw an exception for null frame" in {
      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val thrown = the[IllegalArgumentException] thrownBy LinearRegressionModel.train(null, "label", List("obs1"))
      thrown.getMessage should equal("requirement failed: frame is required")
    }

    "throw an exception for null frame in predict" in {

      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val model = LinearRegressionModel.train(frame, "label", List("obs1"))

      val thrown = the[IllegalArgumentException] thrownBy model.predict(null, None)
      thrown.getMessage should equal("requirement failed: require frame to predict")
    }

    "throw an exception for invalid test column" in {

      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val model = LinearRegressionModel.train(frame, "label", List("obs1"))

      val thrown = the[IllegalArgumentException] thrownBy model.test(null, "label", Some(List("obs1", "obs2")))
      thrown.getMessage should equal("requirement failed: Number of columns for train and test should be same")
    }

    "throw an exception for null value column" in {
      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val thrown = the[IllegalArgumentException] thrownBy LinearRegressionModel.train(frame, null, List("obs1"))
      thrown.getMessage should equal("requirement failed: valueColumn must not be null nor empty")
    }

    "throw an exception for null observation column" in {
      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val thrown = the[IllegalArgumentException] thrownBy LinearRegressionModel.train(frame, "label", null)
      thrown.getMessage should equal("requirement failed: observationColumn must not be null nor empty")
    }

    "throw an exception for max iterations < 0" in {
      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val thrown = the[IllegalArgumentException] thrownBy LinearRegressionModel.train(frame, "label", List("obs1"), maxIterations = (-1))
      thrown.getMessage should equal("requirement failed: numIterations must be a positive value")
    }

    "throw an exception for reg param < 0" in {
      val rdd = sparkContext.parallelize(labeledPoint)
      val frame = new Frame(rdd, schema)
      val thrown = the[IllegalArgumentException] thrownBy LinearRegressionModel.train(frame, "label", List("obs1"), regParam = (-1.0))
      thrown.getMessage should equal("requirement failed: regParam should be greater than or equal to 0")
    }

  }

}
